clang/lib/DPCT/UserDefinedRules/UserDefinedRules.cpp

```diff
@@ -344,17 +344,79 @@ MetaRuleObject::PatternRewriter::PatternRewriter(
   Subrules = S;
 }
 
+// Read a YAML file recursively and substitute any "!include <filename>"
+// directive with the contents of the referenced file.
+std::unique_ptr<llvm::MemoryBuffer>
+readYAMLFile(const llvm::StringRef &ruleFilePath) {
+  // Load the rule file into a MemoryBuffer
+  llvm::ErrorOr<std::unique_ptr<llvm::MemoryBuffer>> Buffer =
+      llvm::MemoryBuffer::getFile(ruleFilePath);
+  if (!Buffer) {
+    llvm::errs() << "Error: failed to read " << ruleFilePath << ": "
+                 << Buffer.getError().message() << "\n";
+    clang::dpct::ShowStatus(MigrationErrorInvalidRuleFilePath);
+    dpctExit(MigrationErrorInvalidRuleFilePath);
+  }
+
+  // Get the directory path of the rule file.
+  llvm::SmallString<128> directoryPath(ruleFilePath);
+  llvm::sys::path::remove_filename(directoryPath);
+
+  // Split the file content into lines using newline as the delimiter.
+  llvm::StringRef buffContent = std::move(*Buffer)->getBuffer();
+  llvm::SmallVector<StringRef, 16> lines;
+  buffContent.split(lines, "\n");
```

Comment:
Using '\n'  to split the file,  we should make sure the content of the YAML file is ended with '\n' for each line.

Reply:
Do you mean, we should handle '\r\n' for file endings on Win?

Reply:
I mean the YAML file can be user-defined, so the format of the YAML file provided by users could be ended with '\r\n' in each line, so the code should cover this scenario.

Reply:
Updated the logic to handle '\r'.
Please let me know if that is the expected change.

Otherwise, please provide a simple failing case for the scenario.